cleverhans/attacks_tf.py

```diff
@@ -1924,35 +1924,58 @@ def margin_logit_loss(model_logits, label, num_classes=10):
   return loss
 
 
-def _apply_transformation(x, dx, dy, angle, batch_size):
+def _apply_black_border(x, border_size):
+  orig_height = x.get_shape().as_list()[1]
+  orig_width = x.get_shape().as_list()[2]
+  x = tf.image.resize_images(x, (orig_width - 2*border_size,
+                                 orig_height - 2*border_size))
+
+  return tf.pad(x, [[0, 0],
+                 [border_size, border_size],
+                 [border_size, border_size],
+                 [0, 0]], 'CONSTANT')
+
+
+def _apply_transformation(inputs):
+  x, trans = inputs[0], inputs[1]
+  dx, dy, angle = trans[0], trans[1], trans[2]
+
   # Map a transformation onto the input
   angle *= np.pi / 180
   height = x.get_shape().as_list()[1]
   width = x.get_shape().as_list()[2]
-  M = np.array([1, 0, -dx*height,
-                0, 1, -dy*width, 0, 0] * batch_size, dtype=np.float32)
-  theta = tf.constant(M, shape=(batch_size, 8))
 
-  # Pad the image to prevent two-step rotation / translation
-  x = tf.pad(x, [[0, 0], [height, height], [width, width], [0, 0]],
+  M = [1, 0, -dx * height,  # Height
+       0, 1, -dy * width,  # Width
+       0, 0]
+  theta = tf.convert_to_tensor(M)
+  theta = tf.expand_dims(theta, axis=0)
+
+  # Pad the image to prevent two-step rotation / translation from truncating corners
+  max_dist_from_center = (float(np.max([height, width])) * np.sqrt(2)) / 2
+  min_edge_from_center = float(np.min([height, width])) / 2
+  padding = np.ceil(max_dist_from_center - min_edge_from_center).astype(np.int32)
+  x = tf.pad(x, [[0, 0],
+                 [padding, padding],
+                 [padding, padding],
+                 [0, 0]],
              'CONSTANT')
+
   # Rotate and translate the image
   x = tf.contrib.image.rotate(x, angle, interpolation='BILINEAR')
   x = tf.contrib.image.transform(x, theta, interpolation='BILINEAR')
-
   return tf.image.resize_image_with_crop_or_pad(x, height, width)
 
 
-def spm(x, model, batch_size=128, y=None, n_samples=None, dx_min=-0.1,
+def spm(x, model, y=None, n_samples=None, dx_min=-0.1,
         dx_max=0.1, n_dxs=5, dy_min=-0.1, dy_max=0.1, n_dys=5,
-        angle_min=-30, angle_max=30, n_angles=11):
+        angle_min=-30, angle_max=30, n_angles=11, black_border_size=0):
```

Comment:
Should we change the default `n_angles` to 31 (same as original paper)?

Reply:
Updated!